### ISSUE TITLE
Python tests need to depend on bindings

### ISSUE THREAD

**Issue by nilsdeppe** [sxs-collaboration/spectre#2631]
# Bug reports:

Some of the python tests, like `GenerateXdmf` depend on having python bindings in order to call `unit_test_path` in the informer module. This is used to avoid writing test files into the source directory.

As far as I can tell we have a few options:
1. promote the python bindings to being "first class citizens" in that they are always built if there is a python executable and the python libs available (i.e. require pybind11 if core python is available). This should work even on HPC systems, but we should retain the ability to completely disable python.
2. we write a "manual" binding (using the raw python libs ourselves instead of pybind11) to wrap the `unit_test_path` (and other functions) needed by possibly all python tests.
3. we use CMake's `configure_file` to copy python tests to the build directory.
4. we use CMake/CTest to pass the build path to the python unit tests. E.g. `python TestBlah.py --unit-test-path`

I think 4 would be the safest way to go if it's doable.

@nilsleiffischer what are your thoughts?

**Comment by nilsvu** [comment]
I think the way #1578 works around this, i.e. by using `spectre_add_python_bindings_test` instead of `spectre_add_python_test`, is a pretty good solution already - all it does is skip the test when no Pybindings are built. Unless we need to run those Python tests that depend on the `unit_test_path` but the code they're testing doesn't depend on the Pybindings, on systems where we don't build the Pybindings, we don't need to do anything. In other words: Instead of your option 4 we could just say: If you need to run those tests, enable Pybindings. If you feel like we absolutely have to support running those tests on systems where we can't build Pybindings then I think your option 4 is the only solution. I think your option 1 basically means we just make building pybindings opt-out instead of opt-in.

**Comment by nilsdeppe** [comment]
1 isn't really a solution because, unless it either fixed itself or you fixed it, we can't run pybindings with shared libraries enabled. If that is resolved then I think we can effectively require the bindings. Disabling a test for `X` because of, what is essentially a bug in our build system (or code if you count the shared libs+bindings issue), isn't great. I've had a few times where the python tests (either bindings or not) have failed in CI but since I build with shared libs locally (I don't want all my disk space going to spectre), I can't really catch this easily. I should be able to change things like `GenerateXdmf` and have the tests be useful without depending on the (not fully functional) bindings.

For 4., is it possible/easy to pass in an argument to the python tests and have the `unittest` class set that up? Maybe more specifically, how much code would this add to each python test?

**Comment by nilsvu** [comment]
Passing the `unit_test_path` to the Python test wouldn't be too bad, it's basically this code: https://github.com/sxs-collaboration/spectre/pull/2609#discussion_r534376948

If the reason we can't fully use Pybindings is that they don't work with shared libs then we should probably get that working though. There's also other things that break on my machine when trying to build with shared libs, mostly related to linking with Charm, so I got stuck on that when trying to get shared libs working. Perhaps we should make solid support for shared libs a priority then.

**Comment by nilsdeppe** [comment]
Oh okay, yea that would work. Requires adding a bit of code to each non-bindings test, but not that bad.

I think it would be good to make shared libs fully functional. Building with shared libs also makes interacting with certain profiles a _lot_ easier. With static libs the profilers actually need to be handled in our build systems, while with shared libs we can just do `profiler ./bin/Executable --input-file...` (a lot easier).

**Comment by nilsvu** [comment]
Ok, I think the main thing we'll need to do before shared libs can work reliably is to link Charm as a standard target in cmake, at least that's where the issues on my machine came in. Do you have an idea how to do that? I remember you mentioned we may be able to remove some of the linker wrappers with the new version of Charm? I think that would make things a lot more robust, also when we try to link an external charm into the container.

**Comment by nilsdeppe** [comment]
Yep, `charmc -print-building-blocks` prints out all the flags. We then need to parse them out (shouldn't be difficult) and set the appropriate flags.

**Comment by nilsdeppe** [comment]
#2669 enables shared libs support. Once that's in, @nilsleiffischer would you mind adding the capability for python tests to get the `unit_test_path` and the build directory path without needing bindings?
